source/dotnet/Library/AdaptiveCards.Templating/AdaptiveCardsTemplateVisitor.cs

```diff
@@ -0,0 +1,710 @@
+            /// <param name="jtoken">new data to kept as data context</param>
+            /// <param name="rootDataContext">root data context</param>
+            public DataContext (JToken jtoken, JToken rootDataContext)
+            {
+                token = jtoken;
+                RootDataContext = rootDataContext;
+
+                if (jtoken is JArray)
+                {
+                    IsArrayType = true;
+                }
+                else
+                {
+                    AddDataContextWithKey(dataKeyword, token, token.DeepClone());
+                    AddDataContextWithKey(rootKeyword, token, RootDataContext);
+                }
+            }
+
+            /// <summary>
+            /// overload contructor that takes <paramref name="text"/> which is <c>string</c>
+            /// </summary>
+            /// <exception cref="JsonException"><c>JToken.Parse(text)</c> can throw JsonException if <paramref name="text"/> is invalid json</exception>
+            /// <param name="text">json in string</param>
+            /// <param name="rootDataContext">a root data context</param>
+            public DataContext(string text, JToken rootDataContext) : this(JToken.Parse(text), rootDataContext)
+            {
+            }
+
+            /// <summary>
+            /// Adds <paramref name="jtoken"/> to <paramref name="targetJObj"/> with <paramref name="key"/>
+            /// </summary>
+            /// <param name="key"></param>
+            /// <param name="targetJObj"></param>
+            /// <param name="jtoken"></param>
+            public static void AddDataContextWithKey(string key, JToken targetJObj, JToken jtoken)
+            {
+                if (key == null || targetJObj == null || jtoken == null)
+                {
+                    return;
+                }
+
+                targetJObj[key] = jtoken; 
+            }
+
+            /// <summary>
+            /// retrieve a <see cref="JObject"/> from this DataContext instance if <see cref="JToken"/> is a <see cref="JArray"/> at <paramref name="index"/>
+            /// </summary>
+            /// <param name="index"></param>
+            /// <returns><see cref="JObject"> at<paramref name="index"/> of a <see cref="JArray"/></returns>
+            public JToken GetDataAtIndex(int index)
+            {
+                if (IsArrayType)
+                {
+                    var jarray = token as JArray;
+                    var jtokenAtIndex = jarray[index];
+                    if (jtokenAtIndex is JObject)
+                    {
+                        var jobject = jtokenAtIndex as JObject;
+                        AddDataContextWithKey(dataKeyword, jobject, jtokenAtIndex.DeepClone());
+                        AddDataContextWithKey(rootKeyword, jobject, RootDataContext);
+                        jobject[indexKeyword] = index;
+                    }
+
+                    return jtokenAtIndex;
+                }
+                return new JObject();
```

Comment:
is there a reasonable case where a caller might ask for data at an index when it's not an array?

Reply:
this method will get called by repeating for array type data context, and index is based on length of the array, so it won't happen.

---
In reply to: [414179450](https://github.com/microsoft/AdaptiveCards/pull/3946#discussion_r414179450) [](ancestors = 414179450)